src/spider.js

```diff
@@ -0,0 +1,268 @@
+const WELCOME_TEXT = 'This is a BTP server, please upgrade to WebSockets.'
+const LE_ROOT = '~/letsencrypt'
+const HTTP_REDIRECT_PORT = 80
+const HTTPS_PORT = 443
+
+const PEER_TYPE_DOWNSTREAM = 'downstream'
+const PEER_TYPE_UPSTREAM = 'upstream'
+
+const CONNECT_RETRY_INTERVAL = 1000
+const SEND_RETRY_INTERVAL = 100
+
+// This function starts a TLS webserver on HTTPS_PORT, with on-the-fly LetsEncrypt cert registration.
+// It also starts a redirect server on HTTP_REDIRECT_PORT, which GreenLock uses for the ACME challenge.
+// Certificates and temporary files are stored in LE_ROOT
+function getLetsEncryptServers (domain) {
+  let httpServer
+  const le = LE.create({
+    // server: 'staging',
+    server: 'https://acme-v01.api.letsencrypt.org/directory',
+    acme: leAcmeCore.ACME.create(),
+    store: leStoreBot.create({ configDir: LE_ROOT + '/etc', webrootPath: LE_ROOT + '/var/:hostname' }),
+    challenges: { 'http-01': leChallengeFs.create({ webrootPath: LE_ROOT + '/var/:hostname' }) },
+    agreeToTerms: function (tosUrl, cb) { cb(null, tosUrl) },
+    debug: true
+  })
+  return new Promise((resolve, reject) => {
+    httpServer = http.createServer(le.middleware())
+    httpServer.listen(HTTP_REDIRECT_PORT, (err) => {
+      if (err) { reject(err) } else { resolve() }
+    })
+  }).then(() => {
+    return le.core.certificates.getAsync({
+      email: `letsencrypt+${domain}@gmail.com`,
+      domains: [ domain ]
+    })
+  }).then(function (certs) {
+    if (!certs) {
+      throw new Error('Should have acquired certificate for domains.')
+    }
+    return new Promise((resolve, reject) => {
+      const httpsServer = https.createServer({
+        key: certs.privkey,
+        cert: certs.cert,
+        ca: certs.chain
+      }, (req, res) => {
+        res.end(WELCOME_TEXT)
+      })
+      httpsServer.listen(HTTPS_PORT, (err) => {
+        if (err) { reject(err) } else { resolve([ httpsServer, httpServer ]) }
+      })
+    })
+  })
+}
+
+function BtpSpider (config, connectHandler, msgHandler) {
+  this.peers = {}
+  this.serversToClose = []
+  this.config = config
+  this.connectHandler = connectHandler
+  this.msgHandler = msgHandler
+  this.incarnations = {}
+  // this.myBaseUrl
+}
+BtpSpider.prototype = {
+  maybeListen () {
+    return Promise.resolve().then(() => {
+      // case 1: use LetsEncrypt => [https, http]
+      if (this.config.tls) {
+        this.myBaseUrl = 'wss://' + this.config.tls
+        return getLetsEncryptServers(this.config.tls)
+      }
+
+      // case 2: don't run a server => []
+      if (typeof this.config.listen !== 'number') {
+        return []
+      }
+
+      // case 3: listen without TLS on a port => [http]
+      this.myBaseUrl = 'ws://localhost:' + this.config.listen
+      const server = http.createServer((req, res) => {
+        res.end(WELCOME_TEXT)
+      })
+      return new Promise(resolve => server.listen(this.config.listen, resolve([ server ])))
+    }).then(servers => {
```

Comment:
![image](https://user-images.githubusercontent.com/3262610/30547576-fb642bf0-9c8f-11e7-8820-9cbb529cdc1b.png)

Reply:
We could have a long discussion about that :)

You're right though the promise construction in that function was a bit weird, I [split out the getServers part](https://github.com/interledgerjs/btp-toolbox/pull/1/commits/884ce3dab1d16de6968a1a9af595e0e4a5726162)